kernel/enter.c

```diff
@@ -860,47 +862,78 @@ int myst_enter_kernel(myst_kernel_args_t* args)
             thread->exit_kstack = NULL;
         }
 
-        /* free all non-process threads, waiting for all other threads to
-         * shutdown at the same time. Our thread has not been marked as a zombie
-         * yet. */
+        /* Wait for all child threads to shutdown */
         {
-            myst_thread_t* t = thread->group_next;
-            while (t)
+            myst_assume(thread->group_prev == NULL);
+            while (thread->group_next)
             {
-                myst_thread_t* next = t->group_next;
-                if (t != thread)
-                {
-                    if (t->status != MYST_ZOMBIE)
-                    {
-                        // We still have a thread that has not shut down
-                        // properly yet
-                        myst_sleep_msec(10);
-                        continue;
-                    }
-                    if (t->group_prev)
-                        t->group_prev->group_next = t->group_next;
-                    if (t->group_next)
-                        t->group_next->group_prev = t->group_prev;
-                    myst_signal_free_siginfos(t);
-                    free(t);
-                }
-                t = next;
+                myst_sleep_msec(10);
             }
         }
 
         /* now all the threads have shutdown we can retrieve the exit status */
         exit_status = thread->exit_status;
 
-        if (args->shell_mode)
-            myst_start_shell("\nMystikos shell (exit)\n");
-
         /* release the fdtable */
         if (thread->fdtable)
         {
             myst_fdtable_free(thread->fdtable);
             thread->fdtable = NULL;
         }
 
+        /* Remove ourself from /proc/<pid> so other processes know we have gone
+         * if they check */
+        procfs_pid_cleanup(thread->pid);
+
+        /* Send SIGHUP to all other active processes */
+        myst_send_sighup_child_processes(thread);
+
+        /* Put the thread on the zombie list, although no one will be doing a
+         * waitpid on this top-level process, but it will make it consistent */
+        //        myst_zombify_thread(thread);
```

Comment:
Where does the thread get zombified now?

Reply:
this is the initial process thread and you removed that code. I'll remove this line